Rewrite of the specification of DNSSEC10

matsduf opened the pull request:
## Purpose

There are several issues concerning problems with both the specification and implementation if DNSSEC10:
* #1153 
* #1101 
* #1048 
* https://github.com/zonemaster/zonemaster-engine/issues/1253

The old version of the test case tries to test NSEC/NSEC3 by creating a non-existing domain name, which is quite complex when the zone has a wildcard. NSEC3 opt out increases the complexity.

This rewrite changes the way NSEC/NSEC3 is tested. NSEC/NSEC3 records have two purposes in DNSSEC:

* Providing proof that a certain name does not exist.
* If the name exists, providing proof that a certain record type does not exist in node of that name.

The current version focuses on non-exiting domain names. The new version provided in this PR focuses instead on record type that do not exist, and  it queries for record types that cannot exist in the zone. This will make the implementation less complex.

## Context

The issues above.

Test zones will be added to this PR, but the specification is complete and ready for review.

## Changes

The rewrite of the specification requires a rewrite of the implementation.

## How to test this PR

This has to be reviewed.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -53,37 +56,45 @@ This test case is only relevant if the zone has been DNSSEC signed.
+DS10_NSEC_QUERY_GIVES_ERR_ANSWER   | ERROR   | ns_ip_list | Unexpected DNS record in the answer section on an NSEC query. Fetched from nameservers with IP addresses "{ns_ip_list}".
+DS10_NSEC_QUERY_RESPONSE_ERROR     | ERROR   | ns_ip_list | No response or error in response on query for NSEC. Fetched from the nameservers with IP addresses "{ns_ip_list}".
+DS10_NSEC_RRSIG_VERIFY_ERROR       | ERROR   | ns_ip_list | The signatures (RRSIG with tag {keytag}) for the NSEC record cannot be verified. Fetched from the nameservers with IP addresses "{ns_ip_list}".
+DS10_SERVER_NO_DNSSEC_SUPPORT      | ERROR   | ns_ip_list | The name servers listed here do not support DNSSEC or has not properly configured. Testing for NSEC and NSEC3 has been skipped on these servers. Fetched from the nameservers with IP addresses "{ns_ip_list}".
```
MSIG updated.

I have changed to `ns_list`. Why here and not in every place?

matsduf replied:
In general, I think `ns` and `ns_list` to be more helpful in the output. They require a little more specification if we are strict. Usually we want to loop over the IP addresses, and then what should be reported if the there are two names to the same IP address? Just report one of them (kind of randomly)?

How much extra will the implementation require to output `ns` (`ns_list`) instead of `ns_ip` (`ns_ip_list`)?

One decision could be to say that in the specification the name server name could be implicit except for that in the `msgid` it must be spelled out if it should be `ns` (`ns_list`) or `ns_ip` (`ns_ip_list`).

matsduf replied:
@tgreenx, would it then be OK to say that in the specification it normally only refer to the name server IP except in the msgid where it should be explicit if it should be `{ns}` or `{ns_ip}`? If we think that would be good, then that could be spelled out in the template.

matsduf replied:
Let us try that. See updates.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +106,302 @@ message. The argument names are defined in the [argument list].
+    20. Name server IP address ("NSEC3 NODATA Missing SOA").
+    21. Name server IP address ("NSEC3 RRSIG Verify Error").
+    22. Name server IP address ("NSEC3PARAM In Answer").
+    23. Name server IP address ("NSEC3PARAM QUERY Gives Erroneous Answer").
```
Updated.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +106,302 @@ message. The argument names are defined in the [argument list].
+    4. Else, add name server IP to the *Responds with DNSKEY* set and retrieve
+       the DNSKEY records from the answer section to be used in validation below.
+    5. Send *NSEC* query to the name server IP and do:
+       1. If at least one of the following createria is met, then add the name
```
Updated.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +106,302 @@ message. The argument names are defined in the [argument list].
+                            appointed.
+
+    6. Send *NSEC3PARAM* query to the name server IP and do:
+       1. If at least one of the following createria is met, then add the name
```
Updated.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +106,302 @@ message. The argument names are defined in the [argument list].
+                add name server IP to the *Erroneous Multiple NSEC* set.
+             5. Else do:
+                1. If the owner name of the NSEC record is not *Child Zone* then
+                   add add name server IP to the *NSEC Mismatches Apex* set.
```
Updated (2 places)

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +106,302 @@ message. The argument names are defined in the [argument list].
-10. If both the *Has NSEC* and the *Has NSEC3* are non-empty, then output
-    *[DS10_INCONSISTENT_NSEC_NSEC3]* with the name server IP addresses from the
-    sets (two sets of name server IP addresses).
+14. Create a list of the name server IP the *NSEC3PARAM In Answer* set or in the
```
Updated.

matsduf commented:
> * The section "Terminology" can be removed.

I think it is better to keep it, but with a comment.

>     * In section "Special procedural requirements", use [the proper sentence from the template](https://github.com/zonemaster/zonemaster/blob/c7c6793e8cb7dec1ca9ede5881f622f5cfae9148/docs/internal/templates/specifications/tests/Template01.md?plain=1#L195-L197).

Fixed.

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +107,301 @@ message. The argument names are defined in the [argument list].
+          2. The [RCODE Name] in the response is not "NoError".
+          3. The AA flag is not set in the response.
+       2. Else if the answer section is non-empty, then do:
+          1. Else if the answer section has a NSEC3PARAM RR then add the name
```
```suggestion
          1. If the answer section has a NSEC3PARAM RR then add the name
```

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +107,301 @@ message. The argument names are defined in the [argument list].
+             server IP to the *Expected NSEC Missing* set.
+          2. Else do:
+             1. Add the name server IP to the *NSEC3PARAM Query Gives NSEC NODATA* set.
+             2. If the SOA record is missing the authority section then add name
+                server IP to the *NSEC NODATA Missing SOA* set.
```
```suggestion
             2. If the SOA record is missing the authority section then add the
                name server IP to the *NSEC NODATA Missing SOA* set.
```

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +107,301 @@ message. The argument names are defined in the [argument list].
+             1. Add the name server IP to the *NSEC3PARAM Query Gives NSEC NODATA* set.
+             2. If the SOA record is missing the authority section then add name
+                server IP to the *NSEC NODATA Missing SOA* set.
+             3. Else if the owner name of SOA record is is not *Child Zone* then
```
```suggestion
             3. Else if the owner name of the SOA record is not *Child Zone* then
```

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +107,301 @@ message. The argument names are defined in the [argument list].
+          2. Else, then add the name server IP to the
+             *NSEC3PARAM Query Gives Erroneous Answer* set.
+       3. Else if the answer section is empty, then do:
+          1. If the autority section contains no NSEC record then add the name
```
```suggestion
          1. If the authority section contains no NSEC record then add the name
```

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +107,301 @@ message. The argument names are defined in the [argument list].
+             5. Else do:
+                1. If the owner name of the NSEC record is not *Child Zone* then
+                   add name server IP to the *NSEC Mismatches Apex* set.
+                2. Else if the type list in the NSEC record does not match the
+                   following criteria then add name server IP to the
+                   *NSEC Incorrect Type List* set:
+                   1. Contains SOA, NS, DNSKEY, NSEC and RRSIG.
+                   2. Does not contain NSEC3PARAM or NSEC3.
```
Again: the list must contain SOA, NS, DNSKEY, NSEC and RRSIG but neither NSEC3PARAM nor NSEC3?
I’m thinking that it might be clearer to word this paragraph as:

> Else if the type list in the NSEC record matches at least one of the following criteria then add name server IP to the *NSEC Incorrect Type List* set:
> 1. SOA, NS, DNSKEY, NSEC or RRSIG is missing from the list;
> 2. The list contains NSEC3PARAM or NSEC3.

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -303,6 +433,9 @@ None.
+No special terminology for this Test Case.
+
+
 ### The Non-Existent Query Name
 
 The term "The Non-Existent Query Name" is used for a name in the *Child Zone*,
```
This paragraph is no longer needed, it seems.

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -312,35 +445,52 @@ should not be used as of [RFC 5890][RFC 5890#section-2.3.1], section 2.3.1.
 [RFC 5155#section-3]:                         https://datatracker.ietf.org/doc/html/rfc5155#section-3
+[RFC 5155#section-4]:                         https://datatracker.ietf.org/doc/html/rfc5155#section-4
 [RFC 5155#section-7.2]:                       https://datatracker.ietf.org/doc/html/rfc5155#section-7.2
 [RFC 5890#section-2.3.1]:                     https://datatracker.ietf.org/doc/html/rfc5890#section-2.3.1
```
This link is no longer needed and can be removed.

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -201,11 +201,12 @@ A complete list of all DNS Resource Record types can be found in the
                    *NSEC3 Mismatches Apex* set.
-                2. Else if the type list in the NSEC3 record does not match the
-                   following criteria then add name server IP to the
+                2. Else if the type list in the NSEC3 record matche at least one
```
```suggestion
                2. Else if the type list in the NSEC3 record matches at least one
```

marc-vanderwal reviewed (commented):
Just one more typo :-)

matsduf commented:
@marc-vanderwal, see my response in https://github.com/zonemaster/zonemaster/pull/1179#discussion_r1273413512. Any comment?

@marc-vanderwal, see my response in https://github.com/zonemaster/zonemaster/pull/1179#discussion_r1273441530. Any comment?

@tgreenx, see my response in https://github.com/zonemaster/zonemaster/pull/1179#discussion_r1269237580. Any comment?

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -53,37 +56,46 @@ This test case is only relevant if the zone has been DNSSEC signed.
-DS10_NSEC_MISSING_SIGNATURE        | ERROR   | ns_ip_list | Missing signatures for NSEC record or records.
-DS10_NSEC_RRSIG_VERIFY_ERROR       | ERROR   | ns_ip_list | The signature or signatures on the NSEC record or records cannot be correctly verfied.
-DS10_UNSIGNED_ANSWER               | ERROR   | ns_ip_list, domain, rrtype | RRSIG is missing for record or records in the answer section.
+DS10_ALGO_NOT_SUPPORTED_BY_ZM      | NOTICE  | ns_ip_list, algo_mnemo, algo_num, keytag | DNSKEY with tag {keytag} uses unsupported algorithm {algo_num} ({algo_mnemo}) by this installation of Zonemaster. Fetched from the nameservers with IP addresses "{ns_ip_list}".
```
I agree with that.

marc-vanderwal commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -53,37 +56,46 @@ This test case is only relevant if the zone has been DNSSEC signed.
+DS10_INCONSISTENT_NSEC3            | ERROR   | ns_ip_list | Inconsistent responses from zone with NSEC3. Fetched from nameservers with IP addresses "{ns_ip_list}".
+DS10_INCONSISTENT_NSEC_NSEC3       | ERROR   |ns_ip_list_nsec, ns_ip_list_nsec3| The zone is inconsistent on NSEC and NSEC3. NSEC is fetched from nameservers with IP addresses "{ns_ip_list_nsec}". NSEC3 is fetched from nameservers with IP addresses "{ns_ip_list_nsec3}".
+DS10_MISSING_NSEC_NSEC3            | ERROR   | ns_ip_list | NSEC or NSEC3 is expected but both are missing. Fetched from nameservers with IP addresses "{ns_ip_list}".
+DS10_MIXED_NSEC_NSEC3              | ERROR   | ns_ip_list | Both NSEC and NSEC3 are responded from the zone, which is . Fetched from the nameservers with IP addresses "{ns_ip_list}".
```
Yes please; I think it can be a useful clarification. As for the msgid, I think it should work too.

matsduf commented:
The added reference to DNSSEC05 should be reviewed in the context of PR #1183.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -95,183 +107,302 @@ message. The argument names are defined in the [argument list].
+    *[DS10_NSEC3_NODATA_MISSING_SOA]* with the list of name server IP in the set.
+
+26. If the *NSEC3PARAM Query Gives Erroneous Answer* set is non-empty, then
+    output *[DS10_NSEC3PARAM_Q_GIVES_ERR_ANSWER]* with the list of name server IP
```
Fixed.

matsduf commented:
@tgreenx and @marc-vanderwal, please re-review.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -81,16 +81,16 @@ DS10_NO_DNSSEC_SUPPORT             | NOTICE  | ns_list    | The zone is not DNSS
 DS10_NSEC3_RRSIG_VERIFY_ERROR      | ERROR   |ns_ip_list, keytag| The RRSIG (signature) with tag {keytag} for the NSEC3 record cannot be verified. Fetched from the nameservers with IP addresses "{ns_ip_list}".
 DS10_NSEC_ERR_TYPE_LIST            | ERROR   | ns_ip_list | NSEC record for the zone apex with incorrect type list. Fetched from nameservers with IP addresses "{ns_ip_list}".
-DS10_NSEC_MISMATCH_APEX            | ERROR   | ns_ip_list | NSEC record with a non-apex owner name, which is unexpected. Fetched from nameservers with IP addresses "{ns_ip_list}".
+DS10_NSEC_MISMATCH_APEX            | ERROR   | ns_ip_list | The returned NSEC record has an unexpected non-apex owner name. Fetched from nameservers with IP addresses "{ns_ip_list}".
```
Because of the way NSEC and NSEC3, respectively, works. The NSEC record for apex has apex as owner name. The NSEC3 record for apex has a hash version of the apex name as owner name.

matsduf replied:
The form with verb inflected in singular is directly matching the set. The form with the uninflected verb is a simplified form. I will make both inflected.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -81,16 +81,16 @@ DS10_NO_DNSSEC_SUPPORT             | NOTICE  | ns_list    | The zone is not DNSS
 DS10_NSEC3_MISSING_SIGNATURE       | ERROR   | ns_ip_list | Missing RRSIG (signature) for the NSEC3 record or records. Fetched from the nameservers with IP addresses "{ns_ip_list}".
 DS10_NSEC3_NODATA_MISSING_SOA      | ERROR   | ns_ip_list | Missing SOA record in NODATA response with NSEC3. Fetched from nameservers with IP addresses "{ns_ip_list}".
-DS10_NSEC3_NODATA_WRONG_SOA        | ERROR   | ns_ip_list | Wrong owner name on SOA record in NODATA response with NSEC3. Fetched from nameservers with IP addresses "{ns_ip_list}".
+DS10_NSEC3_NODATA_WRONG_SOA        | ERROR   | ns_ip_list, domain | Wrong owner name ({"domain"}) on SOA record in NODATA response with NSEC3. Fetched from nameservers with IP addresses "{ns_ip_list}".
```
Fixed.

matsduf commented:
@tgreenx and @marc-vanderwal, please re-review.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -66,35 +66,35 @@ This test case is only relevant if the zone has been DNSSEC signed.
+DS10_NSEC3_MISMATCHES_APEX         | ERROR   | ns_list | The returned NSEC3 record unexpectedly does not match the zone name. Fetched from nameservers with IP addresses "{ns_list}".
+DS10_NSEC3_MISSING_SIGNATURE       | ERROR   | ns_list | Missing RRSIG (signature) for the NSEC3 record or records. Fetched from the nameservers with IP addresses "{ns_list}".
+DS10_NSEC3_NODATA_MISSING_SOA      | ERROR   | ns_list | Missing SOA record in NODATA response with NSEC3. Fetched from nameservers with IP addresses "{ns_list}".
+DS10_NSEC3_NODATA_WRONG_SOA        | ERROR   | ns_list, domain | Wrong owner name ({"domain}"}) on SOA record in NODATA response with NSEC3. Fetched from nameservers with IP addresses "{ns_list}".
```
Fixed.

matsduf commented:
@tgreenx, I did another change. After changing from `ns_ip_list` to `ns_list` it makes more sense to just refer till "name servers" in msgid, not "name servers with IP address". Please re-review.

matsduf commented:
> A minor thing but now I see that some msgids say "Fetched from the nameservers [...]" and others "Fetched from nameservers [...]" (without "the"). Feel free to harmonize that if you want. I'd go for the latter form.

Updated. I also replaced "nameserver" with "name server" to be consistent. The second form is probably the most accepted.

matsduf commented:
@tgreenx and @marc-vanderwal, please re-review.

matsduf commented:
@tgreenx, are you fine with this? Could you implement it?

matsduf commented:
@tgreenx and @marc-vanderwal, I resolved a minor conflict. Please re-review.

matsduf commented:
@tgreenx, is this something we can go ahead with? Can you review again?

matsduf commented:
> > We should wait for test zones specification (and data) first, so I propose to postpone to v2024.1.

Good point! I will do that.

marc-vanderwal reviewed (commented):
Except for some minor typos, I see no issue.

matsduf commented:
@tgreenx and @marc-vanderwal, please review.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -92,187 +111,401 @@ severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+          1. If the answer section has a NSEC RR then do:
+             1. Add the name server IP to the *NSEC In Answer* set.
+             2. If the owner name of the NSEC record is not *Child Zone* then
+                add name server IP to the *NSEC Mismatches Apex* set.
```
Fixed. I will create another scenario to cover this situation.

marc-vanderwal reviewed (approved):
Looks good to me.

matsduf commented on docs/public/specifications/tests/DNSSEC-TP/dnssec10.md:
```diff
@@ -92,187 +111,399 @@ severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+          1. There is no DNS response.
+          2. The [RCODE Name] in the response is not "NoError".
+          3. The AA flag is not set in the response.
+       2. Else if the answer section is non-empty, then do:
+          1. If the answer section has a NSEC3PARAM RR with owner name
+             *Child Zone* then add the name server IP to the
+             *NSEC3PARAM In Answer* set.
+          2. Else, then add the name server IP to the
+             *NSEC3PARAM Query Gives Erroneous Answer* set.
```
Yes. I updated before for NSEC and not NSEC3PARAM. I think it is better if it is consistent. I will update and then also update the scenarios. I hope that sound OK for you.
